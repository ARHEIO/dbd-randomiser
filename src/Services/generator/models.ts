export enum IRank {
  BROWN = 1,
  YELLOW,
  GREEN,
  PURPLE,
  IRIDESCANT,
  GOLD
}

export interface IBasic {
  name: string;
  icon: string;
  rank: IRank | null;
}

export interface IModifiable extends IBasic {
  upgradables: IBasic[]
}

export type Killer = IModifiable; // doesn't use rank
export type Item = IModifiable;
export type Survivor = IBasic;
export type Addon = IBasic;
export type Perk = IBasic;
