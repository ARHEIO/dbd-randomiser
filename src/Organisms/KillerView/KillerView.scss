.killer-container {
  h2,
  p {
    text-shadow: 1px 1px 10px hsla(0, 0%, 100%, 0.25), -1px -1px 10px hsla(0, 0%, 100%, 0.25);
    color: #FFFFFF;
  }

  .killer-toolbar {
    align-items: center;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    margin: 0 8px 32px;
    margin-bottom: 32px;

    & > *:first-child {
      margin-bottom: 16px;

      @media (min-width: 960px) {
        margin: 0;
      }
    }

    @media (min-width: 960px) {
      flex-direction: row;
      margin: 0 0 32px;
    }
  }
}
