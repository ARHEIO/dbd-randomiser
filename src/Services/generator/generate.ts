import axios from 'axios';
import {
  GeneratedKiller, GeneratedSurvivor,
} from './models';
import config from '../../config';

export const generateKiller = (): Promise<{data: GeneratedKiller}> => axios.get(`${config.endpoints.randomiserApi}/killer`);

export const generateSurvivor = (): Promise<{data: GeneratedSurvivor}> => axios.get(`${config.endpoints.randomiserApi}/survivor`);
