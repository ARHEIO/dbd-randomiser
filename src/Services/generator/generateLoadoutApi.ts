import axios from 'axios';
import { GeneratedKiller, GeneratedSurvivor } from './models';
import config from '../../config';

export const generateKiller = (): Promise<{data: GeneratedKiller}> => axios.get(`${config.endpoints.randomiserApi}/loadout?q=killer`);

export const generateSurvivor = (): Promise<{data: GeneratedSurvivor}> => axios.get(`${config.endpoints.randomiserApi}/loadout?q=survivor`);
