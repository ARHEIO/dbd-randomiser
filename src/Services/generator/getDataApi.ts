import axios from 'axios';
import config from '../../config';
import {
  Killer, Survivor, Item, Perk,
} from './models';

export const getKiller = (id: number): Promise<{data: Killer}> => axios
  .get(`${config.endpoints.randomiserApi}/killer/${id}`);
export const getKillerWithAddons = (id: number): Promise<{data: Killer}> => axios
  .get(`${config.endpoints.randomiserApi}/killer/${id}/?expands=upgradables`);
export const getKillers = (): Promise<{data: Killer[]}> => axios
  .get(`${config.endpoints.randomiserApi}/killer`);

export const getSurvivor = (id: number): Promise<{data: Survivor}> => axios
  .get(`${config.endpoints.randomiserApi}/survivor/${id}`);
export const getSurvivors = (): Promise<{data: Survivor[]}> => axios
  .get(`${config.endpoints.randomiserApi}/survivor`);

export const getItem = (id: number): Promise<{data: Item}> => axios
  .get(`${config.endpoints.randomiserApi}/survivor/${id}`);
export const getItemWithAddons = (id: number): Promise<{data: Item}> => axios
  .get(`${config.endpoints.randomiserApi}/survivor/${id}?expands=upgradables`);
export const getItems = (): Promise<{data: Item[]}> => axios
  .get(`${config.endpoints.randomiserApi}/survivor`);

export const getSurvivorPerk = (id: number): Promise<{data: Perk[]}> => axios
  .get(`${config.endpoints.randomiserApi}/perk/survivor/${id}`);
export const getSurvivorPerks = (): Promise<{data: Perk}> => axios
  .get(`${config.endpoints.randomiserApi}/perk/survivor`);

export const getKillerPerk = (id: number): Promise<{data: Perk}> => axios
  .get(`${config.endpoints.randomiserApi}/perk/killer/${id}`);
export const getKillerPerks = (): Promise<{data: Perk[]}> => axios
  .get(`${config.endpoints.randomiserApi}/perk/killer`);
