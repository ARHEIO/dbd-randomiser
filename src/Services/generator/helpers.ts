export const randomNumberGenerator = (max: number) => {
  return Math.floor(Math.random() * max)
}

export const getMultiple = (arrayOfThings: any[], numberToGet: number) => {
  return arrayOfThings && arrayOfThings.length
    ? arrayOfThings
        .map(x => ({ x, r: Math.random() }))
        .sort((a, b) => a.r - b.r)
        .map(a => a.x)
        .slice(0, numberToGet)
    : [];
}

export const getAddons = (addons: any[]) => getMultiple(addons, 2)

export const getPerks = (perks: any[]) => getMultiple(perks, 4)