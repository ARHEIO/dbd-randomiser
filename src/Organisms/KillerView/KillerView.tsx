import React, { useEffect, useState } from 'react';
import './KillerView.scss';
import { generateKiller, GeneratedKiller } from '../../Services/generator/generate';
import KillerPanel from '../../Molecules/KillerPanel/KillerPanel';
import Spinner from '../../Atoms/Spinner/Spinner';


const KillerView = () => {
  const [killerDetails, setKillerDetails] = useState<GeneratedKiller | null>(null);

  const getNewKiller = () => {
    setKillerDetails(null) // set loading state
    generateKiller().then(killer => setKillerDetails(killer))
  };
  
  useEffect(() => {
    generateKiller().then(killer => setKillerDetails(killer))
  }, []);

  return (
    <div className="killer-container">
      <h2>Killer</h2>
      <button onClick={getNewKiller}>Generate</button>
      { killerDetails && killerDetails.name
        ? <KillerPanel killer={killerDetails} />
        : <Spinner />
      }
    </div>
  );
}

export default KillerView;
