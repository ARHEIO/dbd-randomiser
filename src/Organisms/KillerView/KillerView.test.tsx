import React, { ReactElement } from 'react';
import {
  render, fireEvent, wait, act,
} from '@testing-library/react';
import pretty from 'pretty';
import KillerView from './KillerView';

import * as generators from '../../Services/generator/generateLoadoutApi';

jest.mock('../../Molecules/KillerPanel/KillerPanel', () => (): ReactElement => (<p>KillerPanel</p>));
describe('Killer View', () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should render', () => {
    const { container } = render(<KillerView />);
    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });

  it('should generate a killer loadout', async () => {
    const killerSpy = jest.spyOn(generators, 'generateKiller').mockImplementation(jest.fn(() => Promise.resolve({} as any)));
    render(<KillerView />);
    await wait(() => {
      expect(killerSpy).toHaveBeenCalledTimes(1);
    });
  });

  it('should generate a new killer loadout on button press', async () => {
    const killerSpy = jest.spyOn(generators, 'generateKiller').mockImplementation(jest.fn(() => Promise.resolve({} as any)));
    const { container } = render(<KillerView />);
    await wait(() => {
      expect(killerSpy).toHaveBeenCalledTimes(1);
    });

    const button = container.querySelector('.btn-generate-killer');
    if (button) {
      await act(async () => {
        fireEvent.click(button);
      });
      await wait(() => {
        expect(killerSpy).toHaveBeenCalledTimes(2);
      });
    } else {
      expect(true).toBeFalsy();
    }
  });
});
