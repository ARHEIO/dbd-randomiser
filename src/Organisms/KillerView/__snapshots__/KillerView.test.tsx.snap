// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Killer View should render 1`] = `
"<div class=\\"killer-container\\">
  <h2>Killer</h2>
  <div class=\\"killer-toolbar\\"><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained btn-generate-killer MuiButton-containedPrimary\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiButton-label\\">Generate New Loadout</span><span class=\\"MuiTouchRipple-root\\"></span></button><a class=\\"MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary\\" tabindex=\\"0\\" aria-disabled=\\"false\\" href=\\"#/survivor\\"><span class=\\"MuiButton-label\\">To Survivor Loadout</span><span class=\\"MuiTouchRipple-root\\"></span></a></div>
  <div class=\\"survivor loading-spinner-container\\">
    <div class=\\"loading-spinner\\"><svg class=\\"loading-spinner-rear\\">
        <circle cx=\\"50%\\" cy=\\"50%\\" r=\\"18px\\"></circle>
      </svg><svg class=\\"loading-spinner-front\\">
        <circle cx=\\"50%\\" cy=\\"50%\\" r=\\"18px\\"></circle>
      </svg></div>
  </div>
</div>"
`;
